### PR TITLE
feat: add vllm sidecar auto-registration.

### PR DESCRIPTION
**Pull request by fdvty**
## Description

  Adds a standalone vLLM **sidecar** (`vllm_sidecar/`) that auto-registers a vLLM
  instance into xllm-service's etcd, replacing the manual `register_vllm.sh` step
  from the previous PR. Components:
  - `etcd_registry.py` — etcd gateway client (put/lease/keepalive for the instance key)
  - `health.py` — polls the vLLM `/health` endpoint before registering
  - `meta.py` — builds the `InstanceMetaInfo` JSON (backend_type=vllm, type=DEFAULT)
  - `sidecar.py` — the runtime entrypoint tying them together
  - tests for meta and the sidecar runtime

  ## Related Issues
  
  
  ## Change Type
  - [x] New feature
  - [ ] Bug fix / Perf / Refactor / Docs / Test / Build or CI

  ## Pull Request Checklist
  - [x] Commit/PR title follows `<type>: <subject>`.
  - [x] Rebased onto the latest `main`.
  - [x] No C++ changed; Python formatted (ruff) and unit-tested (15 cases pass).

  ## Reviewer Notes
  - Self-contained new Python package; ~1/3 of the diff is unit tests.
  - Builds on the vLLM backend merged in #68; supersedes the manual
    `register_vllm.sh` registration with automatic etcd registration.

### CODE REVIEW

File: vllm_sidecar/etcd_registry.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    def __init__(
+        self,
+        endpoints: str | list[str],
+        username: str = "",
+        password: str = "",
+        timeout: float = 3.0,
+    ) -> None:
+        # `endpoints` may be a comma-separated string or a list of host:port.
+        if isinstance(endpoints, str):
+            endpoints = [e.strip() for e in endpoints.split(",") if e.strip()]
```
[the remark this answers is not in the document]
Comment (fdvty) [reply vllm_sidecar/etcd_registry.py:60]: Good catch. Fixed — endpoints that already carry http:///https:// are kept as-is now, only bare host:port gets the default http:// prefix. Added a unit test covering all three cases.

File: vllm_sidecar/etcd_registry.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        # etcd v3 gateway expects the token in the Authorization header.
+        self._session.headers["Authorization"] = token
+
+    def _post(self, path: str, body: dict) -> dict:
+        last_err = None
+        for base in self._bases:
+            try:
+                r = self._session.post(base + path, json=body, timeout=self._timeout)
+                if r.status_code == 200:
+                    return r.json()
```
[the remark this answers is not in the document]
Comment (fdvty) [reply vllm_sidecar/etcd_registry.py]: Good catch. A 200 with a non-JSON body (e.g. an HTML error page from a proxy) now records the decode error and falls through to the next endpoint instead of raising a raw ValueError; if all endpoints fail it surfaces as EtcdError. Added a unit test covering the non-JSON 200 case.

File: vllm_sidecar/etcd_registry.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        lease_id = resp.get("ID")
+        if not lease_id:
+            raise EtcdError(f"lease grant returned no ID: {resp}")
+        return lease_id
+
+    def lease_keepalive(self, lease_id: str) -> int:
+        """Refresh a lease once; returns the remaining TTL (0 == lease gone)."""
+        resp = self._post("/v3/lease/keepalive", {"ID": lease_id})
+        ttl = resp.get("result", {}).get("TTL", "0")
+        return int(ttl)
```
[the remark this answers is not in the document]
Comment (fdvty) [reply vllm_sidecar/etcd_registry.py]: Good catch. {"result": null} (proto3 JSON for an empty/expired-lease message) now falls back to TTL 0 via resp.get("result") or {} instead of raising AttributeError. Added a unit test for the null-result case.

File: vllm_sidecar/etcd_registry.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+            "/v3/kv/put", {"key": _b64(key), "value": _b64(value), "lease": lease_id}
+        )
+
+    def get(self, key: str) -> str | None:
+        """Return the string value at ``key`` or None if absent."""
+        resp = self._post("/v3/kv/range", {"key": _b64(key)})
+        kvs = resp.get("kvs")
+        if not kvs:
+            return None
+        return base64.b64decode(kvs[0]["value"]).decode("utf-8")
```
[the remark this answers is not in the document]
Comment (fdvty) [reply vllm_sidecar/etcd_registry.py]: Good catch. An omitted value field (proto3 JSON drops empty values) now decodes to "" via .get("value", "") instead of raising KeyError. Added a unit test for the empty-value case.

File: vllm_sidecar/health.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+            return False
+
+    def served_model(self) -> str | None:
+        """Best-effort first model id from `/v1/models`, for logging only."""
+        try:
+            r = requests.get(self._base + "/v1/models", timeout=self._timeout)
+            data = r.json().get("data", [])
+            return data[0]["id"] if data else None
+        except (requests.RequestException, ValueError, KeyError, IndexError):
+            return None
```
[the remark this answers is not in the document]
Comment (fdvty) [reply vllm_sidecar/health.py:53]: Done — VllmHealthProbe now holds a requests.Session and reuses it for both /health and /v1/models, so the periodic probes share one keep-alive connection instead of opening a socket per poll.

**Comment by fdvty** [comment]
Thanks for the thorough review! Addressed every comment, and proactively audited the whole PR for the same classes of issues:
- etcd_registry.py — hardened the entire JSON-parsing surface against proto3 omit-empty/null: non-object 200 bodies coerce to {}, non-JSON 200 falls through to the next endpoint, lease_keepalive tolerates null result/TTL, get tolerates an omitted value, and endpoint scheme prefixes are preserved.
- health.py — VllmHealthProbe now reuses a single requests.Session (keep-alive) instead of a socket per poll.
- sidecar.py / meta.py / demo — fixed a stale doc comment and removed a heartbeat-only flag (--xllm-service-url) from the demo startup that this PR's sidecar doesn't accept yet. Added unit tests for each new edge case (21 tests total). Suggest Squash and merge to collapse the review fixes into the feature commit.

**Review by DongheJin — APPROVED** [review]

File: vllm_sidecar/sidecar.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
+# Copyright 2025 The xLLM Authors. All Rights Reserved.
```
Comment (XuZhang99) [inline vllm_sidecar/sidecar.py]: 2026

Comment (fdvty) [reply vllm_sidecar/sidecar.py]: Fixed — updated the copyright header to 2026 on all new sidecar files. Thanks!

**Review by XuZhang99 — APPROVED** [review]
